finesse/gui/serial_view.py

```diff
@@ -1,138 +1,206 @@
-
-    def on_baud_rate_changed(self, new_baud_rate: str, device_name: str) -> None:
-        """Callback to deal with a change of baud rate for the given device.
-
-        Args:
-            new_baud_rate: The new port selected.
-            device_name: Name of the device affected.
-        """
-        logging.info(
-            f"{device_name} - old baud_rate: {self._devices[device_name]['baud_rate']} "
-            f"- new baud_rate: {new_baud_rate}"
-        )
-        self._devices[device_name]["port"] = new_baud_rate
+        # Try to load the port from settings
+        if avail_ports:
+            saved_port = cast(
+                str, settings.value(f"{_KEY_PREFIX}/{self.name}/port", avail_ports[0])
+            )
+            self.ports.setCurrentText(saved_port)
+
+        self.baudrates = QComboBox()
+        """The available baudrates for the device."""
+        self.baudrates.addItems([str(br) for br in avail_baudrates])
+        layout.addWidget(self.baudrates, row, 2)
+
+        # Try to load the baudrate from settings, falling back on the device's default
+        if avail_baudrates:
+            saved_baudrate = cast(
+                int,
+                settings.value(
+                    f"{_KEY_PREFIX}/{self.name}/baudrate", device.default_baudrate
+                ),
+            )
+            self.baudrates.setCurrentText(str(saved_baudrate))
+
+        self.open_close_btn = QPushButton("Open")
+        """A button for opening and closing the port manually."""
+        self.open_close_btn.setCheckable(True)
+        self.open_close_btn.clicked.connect(self._on_open_close_clicked)  # type: ignore
+        layout.addWidget(self.open_close_btn, row, 3)
+
+        pub.subscribe(self._set_button_to_close, f"serial.{self.name}.opened")
+        pub.subscribe(self._set_button_to_open, f"serial.{self.name}.close")
+        pub.subscribe(self._show_error_message, f"serial.{self.name}.error")
+
+    def _set_button_to_open(self):
+        """Change the button to say Open."""
+        self.open_close_btn.setText("Open")
+
+    def _set_button_to_close(self):
+        """Change the button to say Close."""
+        self.open_close_btn.setText("Close")
+
+    def _show_error_message(self, error: BaseException) -> None:
+        """Show an error message when something has gone wrong with the device."""
+        QMessageBox(
+            QMessageBox.Icon.Critical,
+            "A device error has occurred",
+            f"A fatal error has occurred with the {self.label} device.",
+        ).exec()
+
+    def _open_device(self) -> None:
+        """Open the specified serial device."""
+        port = self.ports.currentText()
+        baudrate = int(self.baudrates.currentText())
+
+        # Remember these settings for the next time program is run
+        settings.setValue(f"{_KEY_PREFIX}/{self.name}/port", port)
+        settings.setValue(f"{_KEY_PREFIX}/{self.name}/baudrate", baudrate)
```

Comment:
```suggestion
        settings.setValue(f"serial/{self.name}/port", port)
        settings.setValue(f"serial/{self.name}/baudrate", baudrate)
```
Also, I would set these values after trying to connect to the device and being successful. If there's any error, the new settings should not be saved in the settings.

Reply:
Both good points!